flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java

```diff
@@ -403,4 +403,58 @@ public boolean isRunning() {
             return currentTaskRunning;
         }
     }
+
+    private boolean containsPrimaryKey() {
+        return !CollectionUtil.isNullOrEmpty(currentSnapshotSplit.getSplitKeyType().getFields());
+    }
+
+    /** Collect records need to be sent, except low/high watermark. */
+    interface SnapshotRecords {
+        void collect(SourceRecord record, boolean reachBinlogStart);
+
+        Collection<SourceRecord> getRecords();
+    }
+
+    /** Collect records with primary key. May upsert binlog events. */
+    class PrimaryKeySnapshotRecords implements SnapshotRecords {
+        private final Map<Struct, SourceRecord> snapshotRecords = new LinkedHashMap<>();
+
+        @Override
+        public void collect(SourceRecord record, boolean reachBinlogStart) {
+            if (!reachBinlogStart) {
+                snapshotRecords.put((Struct) record.key(), record);
+            } else {
+                if (isRequiredBinlogRecord(record)) {
+                    // upsert binlog events through the record key
+                    upsertBinlog(snapshotRecords, record);
+                }
+            }
+        }
+
+        @Override
+        public Collection<SourceRecord> getRecords() {
+            return snapshotRecords.values();
+        }
+    }
+
+    /** Collect records without primary key. There are no binlog events. */
+    static class NoPrimaryKeySnapshotRecords implements SnapshotRecords {
+        private final List<SourceRecord> snapshotRecords = new LinkedList<>();
+
+        @Override
+        public void collect(SourceRecord record, boolean reachBinlogStart) {
+            if (!reachBinlogStart) {
+                snapshotRecords.add(record);
```

Comment:
How to handle the delete and update events?

Reply:
IMHO, the snapshot stage for the table without primary key only read snapshot data, then binlog data would be read only in the incremental stage. So there are no delete and update events in SnapshotSplitReader if the table has no primary key. Please correct me.

Reply:
It is a good idea to skip the snapshot backfill task when there is a single split. I think you should set this in source but I do not see the relative changes.